packages/lib-classifier/src/store/SubjectStore/SubjectStore.js

```diff
@@ -223,40 +277,111 @@ const SubjectStore = types
+            self.loadingState = asyncStates.loading
+            const subjects = yield _fetchSubjects({ apiUrl, params })
+            self.loadingState = asyncStates.success
+            if (subjects?.length > 0) {
+              self.append(subjects)
+            }
           }
         } catch (error) {
           console.error(error)
           self.loadingState = asyncStates.error
         } 
       }
     }
+    /** Insert new subjects into the queue but maintain priority ordering */
+    function prepend(newSubjects = []) {
+      if (newSubjects.length > 0) {
+        const priorityQueue = Array.from(self.resources.values())
+        priorityQueue.unshift(...newSubjects)
+
+        priorityQueue.forEach(subject => {
+          const subjectSnapshot = subject.toJSON ? subject.toJSON() : subject
+          try {
+            self.resources.delete(subjectSnapshot.id)
+            self.resources.put(subjectSnapshot)
+          } catch (error) {
+            console.error(`Subject ${subject.id} is not a valid subject.`)
+            console.error(error)
+          }
+        })
+      }
+    }
+    /** Get the previous subject, in an indexed, prioritised set, and make it the active subject */
+    function previous() {
+      const root = getRoot(self)
+      const workflow = tryReference(() => root.workflows.active)
+      const activeSubject = tryReference(() => self.active)
+      const currentPriority = activeSubject?.priority
+
+      if (workflow?.hasIndexedSubjects) {
+        self.available.clear()
+        let previousSubjects = []
+        if (activeSubject) {
+          const queue = Array.from(self.resources.values())
+          const activeIndex = queue.indexOf(activeSubject)
+          previousSubjects = queue.slice(0, activeIndex)
+        }
+        if (previousSubjects.length > 0) {
+          const previousSubject = previousSubjects[previousSubjects.length - 1]
+          self.setActiveSubject(previousSubject.id)
+        }
+        if (previousSubjects.length <= MINIMUM_QUEUE_SIZE) {
+          self.buildPreviousQueue(currentPriority)
+        }
+      }
+    }
 
     function reset () {
       self.resources.clear()
       self.available.clear()
     }
 
+    function setActiveSubject(subjectID) {
+      self.active = subjectID
+      if (process.env.NODE_ENV !== 'test') console.log('Loading subject', subjectID)
+    }
+
     function setOnReset(callback) {
       self.onReset = callback
     }
+    /** Shift the subject queue by one subject, so that the active subject is always the first subject. */
+    function shift() {
+      const subject = tryReference(() => self.active)
+
+      if (subject) {
+        self.resources.delete(subject.id)
+      }
+      if (self.resources.size < MINIMUM_QUEUE_SIZE) {
+        console.log('Fetching more subjects')
+        self.populateQueue()
+      }
+      if (self.first) {
+        self.setActiveSubject(self.first.id)
+      }
+    }
```

Comment:
I don't entirely understand this approach. It seems like you're trying to make the map work like an array. Why does the active subject need to be the first in the map? Why does any subject need to be removed from the map?

Reply:
This is copied straight from the code that advances the queue on master. The only change I've made is to add `self.first` as a shorthand for `self.resources.values().next().value`.

Remove the current active subject from the map:
https://github.com/zooniverse/front-end-monorepo/blob/031f7eb3b8beabaff65a934d2dc24785c8fcf2c6/packages/lib-classifier/src/store/SubjectStore/SubjectStore.js#L159-L163

Then make the first subject in the map (`self.resources.values().next().value`) active.
https://github.com/zooniverse/front-end-monorepo/blob/031f7eb3b8beabaff65a934d2dc24785c8fcf2c6/packages/lib-classifier/src/store/SubjectStore/SubjectStore.js#L165-L167

Reply:
I'm trying to make sense now in the context of and consideration what `AvailableSubjects` does. I don't understand why we have two stores that do extremely similar things.

Reply:
`AvailableSubjects` is populated when you press 'Next Available Subject' from a finished subject. eg. imagine a set that has 200 prioritised subjects, with the first unclassified subject being page 100. If you were to use the subject picker to select a subject near the beginning, number 10 say, then subjects 10 to 19 are now queued up in the `subjects.resources` but all are finished. When you press 'Next Available Subject' from any of them, you will jump straight to subject 100, which triggers a rebuild of 'subjects.resources', containing subjects 100 to 109.

If you were to use the Next button, on the other hand, that would advance the subject queue in batches of 10, in order, so that you would have to page through subjects 10 to 99 before reaching the first unclassified subject. `subjects.available` gives you the ability to jump long queues. There are only 10 subjects in each Scarlets & Blues subject set at the moment, but the live project will have ~450 pages per set, with volunteers being able to start transcribing at any point in the set and then move back and forward through the set as they please (using the Next and Previous buttons here.)

Reply:
I do think this is going to be a lot easier to test once the RGBE project has some full, indexed subject sets set up.